Commonmark with value classes

obask opened the pull request:
This PR includes both changes to block nodes as well as inline nodes.

For now the Processor is a noop, but it could be helpful later if we decide to offload some operations from the rendering pipeline. It would be easier to decide once we implement quick typing improvements.

rock3r commented on markdown/core/src/main/kotlin/org/jetbrains/jewel/markdown/rendering/DefaultInlineMarkdownRenderer.kt:
```diff
@@ -30,78 +17,73 @@ public open class DefaultInlineMarkdownRenderer(rendererExtensions: List<Markdow
-        var child = node.firstChild
-
-        while (child != null) {
+    private fun Builder.appendInlineMarkdownFrom(inlineMarkdown: Iterator<InlineMarkdown>, styling: InlinesStyling) {
```
Passing an `Iterator` around is kinda weird, because you can't know its state in advance. I'd prefer if it was receiving an `Iterable` instead so it can own its iterator

obask replied:
Iterator worked until I made things nullable... so I ended up simply using a `List`

rock3r commented on markdown/core/src/main/kotlin/org/jetbrains/jewel/markdown/rendering/DefaultMarkdownBlockRenderer.kt:
```diff
@@ -127,83 +127,18 @@ public open class DefaultMarkdownBlockRenderer(
+            3 -> render(block, styling.h3)
+            4 -> render(block, styling.h4)
+            5 -> render(block, styling.h5)
+            6 -> render(block, styling.h6)
```
Missing the else case

rock3r commented on markdown/core/src/main/kotlin/org/jetbrains/jewel/markdown/rendering/DefaultMarkdownBlockRenderer.kt:
```diff
@@ -373,9 +305,10 @@ public open class DefaultMarkdownBlockRenderer(
         ) {
             Column(Modifier.padding(styling.padding)) {
-                if (block.mimeType != null && styling.infoPosition.verticalAlignment == Alignment.Top) {
+                val mimeType = remember { MimeType.Known.fromMarkdownLanguageName(block.value.info) }
```
Could this be a property or function on our Fenced Block model? Value classes can have properties and functions IIRC. Ideally, I'd like to have less arcane ways to access useful information in cases like this. Our value classes can decorate and enrich the CommonMark models to make it easier to build extensions

rock3r commented on markdown/extension/gfm-alerts/src/main/kotlin/org/jetbrains/jewel/markdown/extensions/github/alerts/GitHubAlertProcessorExtension.kt:
```diff
@@ -40,19 +36,20 @@ public object GitHubAlertProcessorExtension : MarkdownProcessorExtension {
-            }
+        override fun processMarkdownBlock(block: CustomBlock, processor: MarkdownProcessor): MarkdownBlock.CustomBlock? {
+            error("should use CustomNode parser instead")
+//            val children = processor.processChildren(block)
+//
+//            if (children.isEmpty()) return null
+//
+//            return when (block) {
+//                is Caution -> Alert.Caution(children)
+//                is Important -> Alert.Important(children)
+//                is Note -> Alert.Note(children)
+//                is Tip -> Alert.Tip(children)
+//                is Warning -> Alert.Warning(children)
+//                else -> error("Unsupported custom block of type ${block.javaClass.name}")
+//            }
```
Should we remove commented code?

rock3r commented on markdown/extension/gfm-alerts/src/main/kotlin/org/jetbrains/jewel/markdown/extensions/github/alerts/GitHubAlertProcessorExtension.kt:
```diff
@@ -40,19 +36,20 @@ public object GitHubAlertProcessorExtension : MarkdownProcessorExtension {
-                else -> error("Unsupported custom block of type ${block.javaClass.name}")
-            }
+        override fun processMarkdownBlock(block: CustomBlock, processor: MarkdownProcessor): MarkdownBlock.CustomBlock? {
+            error("should use CustomNode parser instead")
```
I don't love APIs that have functions you must override but throw when you call them. Especially if the return type is nullable, wouldn't you just return null? Am I missing something?

rock3r commented on markdown/extension/gfm-alerts/src/test/kotlin/org/jetbrains/jewel/markdown/extensions/github/alerts/GitHubAlertBlockExtensionTest.kt:
```diff
@@ -44,8 +39,8 @@ class GitHubAlertBlockExtensionTest {
         assertTrue(
-            "Parsed node should be a Note node, but was ${firstChild.javaClass.name}",
-            firstChild is Note,
+            "Parsed node should be a Note node, but was $firstChild",
```
I think this message (and all similar following ones) should print firstChild.kind, not the whole firstChild

obask commented:
addressed comments in https://github.com/JetBrains/jewel/commit/76b52827fbae2ebd6b42c8b31ba825c2853c958f but I can't create a chain of PRs, because I can't push to my branches

rock3r commented:
@obask you have write permissions, that shouldn't happen. What does the error say? (we can troubleshoot in chat)
